File: test/2/InsertSort.php
<?php
function insertSort(&$arr) {

    $len = count($arr);

    for($i=1; $i<$len; $i++) {

        $tmp = $arr[$i];
        for ($j = $i-1; $j>=0; $j--) {
            if ($arr[$j] > $tmp) {
                $arr[$j+1] = $arr[$j];
            } else {
                break;
            }
        }
       
        // 条件满足，一定会空出来一个重复的位置为$j+1
        $arr[$j + 1] =$tmp;
    }
}

$nums = [5, 8, 3, 1];
$rs = insertSort($nums);
print_r($nums);

File: sort/InsertSort.go
package main
import "fmt"

func insertSort(arr []int) {
    var length =len(arr)
    for i := 1; i < length; i++ {
        // 用来记录当前的数值
        var tmp int = arr[i]
        var j int = i-1;
        for ;j >= 0; j-- {
            if tmp > arr[j] { // 当前已经是有序的
                break
            }
            arr[j+1] = arr[j] // 交换位置
        }
        // 放入合适的位置，如果循环完整的完成，那么 j 一定是-1，所以这里要+1
        arr[j+1] = tmp
    }
}

func main()  {
	
	var arr []int = []int {6, 7, 1, 3, 2, 5, 4}
	insertSort(arr)
	fmt.Println(arr)

}